### ISSUE TITLE
Adding a custom alert to a payload that already contains a simple alert throws a vague JSONException

### ISSUE THREAD

**Issue by mwkirk** [mwkirk/javapns#73]
_Original author: ajmannat...@gmail.com (October 02, 2011 09:52:54)_

<b>What steps will reproduce the problem?</b>
1. using addCustomAlertBody(&quot;sas&quot;)

<b>What is the expected output? What do you see instead?</b>
Exception in thread &quot;main&quot; org.json.JSONException: JSONObject[&quot;alert&quot;] is not a JSONObject. 

<b>What version of the product are you using? On what operating system?</b>
javapns_2.0_RC.jar

_Original issue: http://code.google.com/p/javapns/issues/detail?id=73_

**Comment by mwkirk** [comment]
_From sype...@gmail.com on October 02, 2011 14:59:50_
Is it possible you already added a simple alert in your payload, before trying to add a custom one?  If you add a simple alert, the payload will link a String to the "alert" key; but if you try to add a custom alert afterwards, it will need to link a JSON object to the same "alert" key, which it won't be able to do because you already added a String to that key.

**Comment by mwkirk** [comment]
_From sype...@gmail.com on October 03, 2011 01:20:11_
Committed 2.0 RC Revision 002 to SVN with an improvement regarding this issue.  If one tries to add a custom alert to a PushNotificationPayload (though any of the addCustomAlert... methods) that already contains a simple alert, a more precise exception is now thrown indicating that a simple alert was already added to the payload (and vice-versa).

**Comment by mwkirk** [comment]
_From sype...@gmail.com on October 03, 2011 02:05:43_
Please ignore revision link in comment above, it got linked automatically by Google Code to an unrelated project entry.

javapns 2.0 RC.002 is available in SVN and not at the link in the comment above.

**Comment by mwkirk** [comment]
_From ajmannat...@gmail.com on October 03, 2011 06:11:04_
This is the code that i am using: 

```
            PushNotificationPayload x = new PushNotificationPayload();

    x.addCustomAlertBody("message body");
    x.addBadge(11);
    x.addSound("default");

    x.addCustomDictionary("entity-id", "4e88438978b8f20c187997c5");
    x.addCustomDictionary("entity-type", "USER");

            Push.payload(x, "cert.p12", "xxxxx", false, "71f87aa.......e");
```

The above code throws the exception.

**Comment by mwkirk** [comment]
_From sype...@gmail.com on October 03, 2011 14:00:41_
I cannot replicate the problem with the code you provided.  I inserted my own keystore, keystore password and device token, and your code successfully pushed a notification to my device.

If that code really throws an exception for you, could you please provide a complete stack trace (with line numbers)?  Thank you.

**Comment by mwkirk** [comment]
_From ajmannat...@gmail.com on October 03, 2011 17:43:51_
I'm still getting the error. Here is the stack trace that i get: 
Exception in thread "main" org.json.JSONException: JSONObject["alert"] not found.
    at org.json.JSONObject.get(JSONObject.java:517)
    at org.json.JSONObject.getJSONObject(JSONObject.java:611)
    at javapns.notification.PushNotificationPayload.getCustomAlert(PushNotificationPayload.java:119)
    at javapns.notification.PushNotificationPayload.addCustomAlertBody(PushNotificationPayload.java:134)
    at Test.main(Test.java:34)

**Comment by mwkirk** [comment]
_From sype...@gmail.com on October 03, 2011 18:46:15_
You do not appear to be using the latest build as suggested in comment 3.  Could you please download the latest build of the library from SVN and try again?  Thanks!

**Comment by mwkirk** [comment]
_From ajmannat...@gmail.com on October 05, 2011 14:56:29_
Sorry, but I am not really sure how to pull that revision through here.

**Comment by mwkirk** [comment]
_From sype...@gmail.com on October 05, 2011 15:39:58_
A shortcut is available on the Requirements wiki page (requirement #1, nightly build), but here is the complete procedure:

1) click on the Source panel in this web site
2) click on Browse
3) navigate down to svn > branches > javapns2
4) click the latest version of the library (javapns_2.0_RC.jar) on the right
5) click "View raw file" in the File info section of the page (your download will start)

**Comment by mwkirk** [comment]
_From ajmannat...@gmail.com on October 05, 2011 16:27:37_
Thanks a lot for your help.

Its working perfectly now. 
Very much appreciated!

**Comment by mwkirk** [comment]
_From sype...@gmail.com on October 05, 2011 16:30:00_
Good news!  Thanks!

Closing as Fixed in 2.0 RC.002.
